core/src/main/scala/kafka/utils/Exit.scala

```diff
@@ -34,20 +34,34 @@ object Exit {
     throw new AssertionError("halt should not return, but it did.")
   }
 
+  def addShutdownHook(runnable: Runnable, name: Option[String] = None): Unit = {
```

Comment:
Should we use a by-name parameter instead?

Reply:
I don't think so because the runnable is dereferenced once, immediately, and the code passes it to the underlying Java implementation `org.apache.kafka.common.utils.Exit`

Reply:
I don't understand why any of that disqualifies a by name parameter. The benefit of using a by-name is that the caller code becomes cleaner (no need for `() => ...`).

Reply:
I got it working with a by-name parameter, and the below test passes.  Is this what you were thinking?  If so, I can add a new commit with the changes for a better look.

```
  @Test
  def shouldNotInvokeShutdownHookImmediately(): Unit = {
    val value = "value"
    val array:Array[Any] = Array(value)

    def sideEffect(): Unit = {
      // mutate the first element
      array(0) = array(0).toString + array(0).toString
    }
    Exit.addShutdownHook(sideEffect) // by-name parameter, not invoked
    // make sure the first element wasn't mutated
    assertEquals(value, array(0))
    Exit.addShutdownHook(sideEffect()) // by-name parameter, not invoked
    // again make sure the first element wasn't mutated
    assertEquals(value, array(0))
    Exit.addShutdownHook(sideEffect, Some("message")) // by-name parameter, not invoked
    // make sure the first element still isn't mutated
    assertEquals(value, array(0))
  }
```

Reply:
@rondagostino yeah